### PR TITLE
Two pairs functionality

### PR DESCRIPTION
**Pull request by harvey-taylor**
Added two pairs test folder
Added Two Pairs Functionality.

### CODE REVIEW

File: PokerGame/PokerEngine/Poker.cs [hunk 1]
```diff
@@ -7,7 +7,13 @@ public class Poker
     {
         public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> handTwo)
         {
-            List<Card> winningHand = CaclulateHandWithBestPair(handOne, handTwo);
+            List<Card> winningHand = CaclulateHandWithBestTwoPair(handOne, handTwo);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: use ```var```

File: PokerGame/PokerEngine/Poker.cs [hunk 2]
```diff
@@ -7,7 +7,13 @@ public class Poker
...
+            if (winningHand != null)
+            {
+                return winningHand;
+            }
+
+            winningHand = CaclulateHandWithBestPair(handOne, handTwo);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: typo: Caclulate

File: PokerGame/PokerEngine/Poker.cs [hunk 3]
```diff
@@ -16,6 +22,53 @@ public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> han
             return CalculateHandWithHighestCardAndKicker(handOne, handTwo);
         }
 
+        private static List<Card> CaclulateHandWithBestTwoPair(List<Card> handOne, List<Card> handTwo)
+        {
+            Rank handOneFirstPair;
+            Rank handOneSeccondPair;
+            bool handOneHasTwoPair = HasTwoPair(handOne, out handOneFirstPair, out handOneSeccondPair);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: use ```var```

File: PokerGame/PokerEngine/Poker.cs [hunk 4]
```diff
@@ -16,6 +22,53 @@ public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> han
...
+            Rank handOneHighestPairHighestCard = CalculateHighestPair(handOneFirstPair, handOneSeccondPair);
+
+            Rank handTwoFirstPair;
+            Rank handTwoSeccondPair;
+            bool handTwoHasTwoPair = HasTwoPair(handTwo, out handTwoFirstPair, out handTwoSeccondPair);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: use ```var```

File: PokerGame/PokerEngine/Poker.cs [hunk 5]
```diff
@@ -16,6 +22,53 @@ public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> han
...
+            List<Card> handOnePairs = new List<Card>()
+            {
+                new Card() { Rank = pairOne },
+                new Card() { Rank = pairTwo }
+            };
+            Rank highestPair = HighestCard(handOnePairs).Rank;
+            return highestPair;
+        }
+
         private static List<Card> CaclulateHandWithBestPair(List<Card> handOne, List<Card> handTwo)
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: more caclulate typo!

File: PokerGame/PokerEngine/Poker.cs [hunk 6]
```diff
@@ -80,6 +133,23 @@ private static bool HasPair(List<Card> hand, out Rank pair)
             return false;
         }
 
+        private static bool HasTwoPair(List<Card> hand, out Rank pairOne, out Rank pairTwo)
+        {
+            var handICanManipulate = new List<Card>(hand);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: rename: cardsWithoutFirstPair

File: PokerGame/PokerEngine/Poker.cs [hunk 7]
```diff
@@ -80,6 +133,23 @@ private static bool HasPair(List<Card> hand, out Rank pair)
...
         }
 
+        private static bool HasTwoPair(List<Card> hand, out Rank pairOne, out Rank pairTwo)
+        {
+            var handICanManipulate = new List<Card>(hand); 
+            bool handHasOnePair = HasPair(hand, out pairOne);
+            Rank ranktoRemove =  pairOne;
+            handICanManipulate.RemoveAll(card => card.Rank == ranktoRemove);
+
+            bool handHasSeccondOnePair = HasPair(handICanManipulate, out pairTwo);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: Typo: Seccond

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace PokerEngineTests
+{
+    [TestClass]
+    public class TwoPairHandTests
+    {
+        [TestMethod]
+        public void hand_with_two_pair_wins_against_a_pair()
+        {
+            List<Card> fivesAndTwosHand = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Jack },
+                new Card() { Rank = Rank.Jack },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Nine }
+            };
+
+            List<Card> actualWinningHand = Poker.CalculateWinningHand(PairOfTwosHand, fivesAndTwosHand);
+
+            List<Card> expectedWinningHand = fivesAndTwosHand;
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            Assert.AreEqual(expectedWinningHand, actualWinningHand);
+        }
+
+        [TestMethod]
+        public void hand_with_two_pair_wins_by_high_card()
+        {
+            List<Card> fivesAndTwosAceHighHand = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Ace },
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Five }
+            };
+
+            List<Card> fivesAndTwosQueenHighHand = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Queen },
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Five }
+            };
+
+            List<Card> actualWinningHand = Poker.CalculateWinningHand(fivesAndTwosQueenHighHand, fivesAndTwosAceHighHand);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+            List<Card> expectedWinningHand = fivesAndTwosAceHighHand;
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+            Assert.AreEqual(expectedWinningHand, actualWinningHand);
+        }
+
+        [TestMethod]
+        public void hand_with_two_pair_wins_by_higher_top_pair()
+        {
+            List<Card> acesAndFivessHand = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Ace },
+                new Card() { Rank = Rank.Ace },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Five }
+            };
+
+            List<Card> fivesAndTwosHand = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Queen },
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Five }
+            };
+
+            List<Card> actualWinningHand = Poker.CalculateWinningHand(acesAndFivessHand, fivesAndTwosHand);
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+            List<Card> expectedWinningHand = acesAndFivessHand;
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngineTest/TwoPairHandTests.cs]: use ```var```

File: PokerGame/PokerEngine/Poker.cs [hunk 18]
```diff
@@ -16,6 +22,53 @@ public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> han
...
+            if (handTwoHasTwoPair)
+            {
+                return handTwo;
+            }
+            return null;
+        }
+
+        private static Rank CalculateHighestPair(Rank pairOne, Rank pairTwo)
+        {
+            List<Card> handOnePairs = new List<Card>()
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: use ```var```

File: PokerGame/PokerEngine/Poker.cs [hunk 19]
```diff
@@ -16,6 +22,53 @@ public static List<Card> CalculateWinningHand(List<Card> handOne, List<Card> han
             return CalculateHandWithHighestCardAndKicker(handOne, handTwo);
         }
 
+        private static List<Card> CaclulateHandWithBestTwoPair(List<Card> handOne, List<Card> handTwo)
+        {
+            Rank handOneFirstPair;
```
Comment (richard-keast-bluefruit) [inline PokerGame/PokerEngine/Poker.cs]: rankOfFirstPair

File: PokerGame/PokerEngine/Poker.cs (hunk 2 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: good spot, cheers

File: PokerGame/PokerEngine/Poker.cs (hunk 1 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: okay cool i think for some reason i was under the impression var didnt work for Lists but im wrong lol

File: PokerGame/PokerEngine/Poker.cs (hunk 3 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: 👍

File: PokerGame/PokerEngine/Poker.cs (hunk 4 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: 👍

File: PokerGame/PokerEngine/Poker.cs (hunk 5 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: 👍

File: PokerGame/PokerEngine/Poker.cs (hunk 6 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: awesome was meant to rename but forgot 👍

File: PokerGame/PokerEngine/Poker.cs (hunk 7 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 8 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            List<Card> fivesAndTwosHand = new List<Card>()
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Ace },
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Five }
+            };
+
+            List<Card> PairOfTwosHand = new List<Card>()
```
[the remark this answers is not in the document]
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+            {
+                new Card() { Rank = Rank.Two },
+                new Card() { Rank = Rank.Jack },
+                new Card() { Rank = Rank.Jack },
+                new Card() { Rank = Rank.Five },
+                new Card() { Rank = Rank.Nine }
+            };
+
+            List<Card> actualWinningHand = Poker.CalculateWinningHand(PairOfTwosHand, fivesAndTwosHand);
```
[the remark this answers is not in the document]
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 9 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 10 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 11 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 12 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 13 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: I bet your sick of Ctrl + V by now 😂

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 14 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 15 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 16 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngineTest/TwoPairHandTests.cs (hunk 17 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngineTest/TwoPairHandTests.cs]: 👍

File: PokerGame/PokerEngine/Poker.cs (hunk 18 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: 👍 also renamed to just hand

File: PokerGame/PokerEngine/Poker.cs (hunk 19 above)
Comment (harvey-taylor) [reply PokerGame/PokerEngine/Poker.cs]: good shout :)

**Comment by richard-keast-bluefruit** [comment]
Looks good for mergin'!

**Pull request merged**